### PR TITLE
Implement the QuinticSpline class

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This PR introduces the `QuinticSpline` class. This class can be seen as an extension of the `iDynTree::CubicSpline` class. The main difference between this class and the `iDynTree` one is the order of the polynomials used in the spline. Using a 5th order spline guarantees to generate a trajectory that minimizes the jerk. (I will add further details for the latest statement later)

### CODE REVIEW

**Comment by prashanthr05** [comment]
Would you point me to a reference document I could review before going through the details of this PR?

**Comment by S-Dafarra** [comment]
> Would you point me to a reference document I could review before going through the details of this PR?

It is probably already clear, but the reference issue is https://github.com/dic-iit/element_dcm-walking/issues/188

File: src/Planners/src/QuinticSpline.cpp
Comment (S-Dafarra) [inline src/Planners/src/QuinticSpline.cpp:693]: The time is going to be elevated at the power of 5. While walking we usually don't reset the time, so this number can grow very big and cause numerical issues. Consider defining a time offset according to the first knot point.

**Comment by prashanthr05** [comment]
> I could not check in detail the mathematical implementation, but I am going to trust the test sweat_smile

I can check the math details when I will do the review. It will be a good exercise for me as well.

File: src/Planners/src/QuinticSpline.cpp
Comment (prashanthr05) [inline src/Planners/src/QuinticSpline.cpp:594]: Could you add a line on how this comes up?

File: src/Planners/src/QuinticSpline.cpp
Comment (prashanthr05) [inline src/Planners/src/QuinticSpline.cpp:611]: this extra line could be removed.

File: src/Planners/src/QuinticSpline.cpp
Comment (prashanthr05) [inline src/Planners/src/QuinticSpline.cpp:421]:
```suggestion
    tripletList.push_back({rowOffset,
                           columnOffset,
                           12 * ( (1 / std::pow(poly[knotIndex - 1].duration, 2)
                                 - (1 / std::pow(poly[knotIndex].duration, 2)) ) });
```

I'd more careful and explicit with the use of braces here. Similarly for other triplet values.

Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp:421]: To be more readable, get a const ref on `poly[knotIndex - 1]` and `poly[knotIndex]` maybe?

File: src/Planners/src/QuinticSpline.cpp
Comment (prashanthr05) [inline src/Planners/src/QuinticSpline.cpp:493]: According to your derivation in https://github.com/dic-iit/element_dcm-walking/issues/188#issuecomment-673542645,  `-2 / std::pow(poly[knotIndex].duration, 2)`  must be raised to the power 3 `-2 / std::pow(poly[knotIndex].duration, 3)`. There seems to be a typo somewhere (most possibly in the issue comment). Please cross-verify.

File: src/Planners/src/QuinticSpline.cpp
Comment (prashanthr05) [inline src/Planners/src/QuinticSpline.cpp:509]: Are you using a `+=` here assuming that the existing contents are zero?

File: src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h
Comment (GiulioRomualdi) [reply src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h:23]: Actually I don't know, I would like to have a similar interface to an [`iDynTree::CubicSpline`](https://robotology.github.io/idyntree/master/classiDynTree_1_1CubicSpline.html). Furthermore I was thinking to add also a function to get the entire trajectory (from the initial to the final time instant) this may help to use this class also with predictive controllers.

Comment (S-Dafarra) [reply src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h:23]: I don't think these things are mutually exclusive. Inheriting from ``Advanceable`` would imply having two additional methods, but you can still use it as you prefer.

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp:685]: Actually `getPositionAtTime` takes as input the current time minus the initial point instant. Let me know if I'm misunderstanding.


https://github.com/dic-iit/bipedal-locomotion-framework/blob/aaa044da923e6ed0dd895b5f293f36e24a5d39ab/src/Planners/src/QuinticSpline.cpp#L782-L784

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: Fixed in 316ad2fc96203dee9d7cc2e34a80e5ea4300b81d

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: Thank you fixed in 316ad2fc96203dee9d7cc2e34a80e5ea4300b81d.

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: Fixed in 316ad2fc96203dee9d7cc2e34a80e5ea4300b81d

File: src/Planners/src/QuinticSpline.cpp
Comment (S-Dafarra) [reply src/Planners/src/QuinticSpline.cpp:685]: Ah great, I did not notice this was a method of ``Impl``. No issue then.

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: I improved the readability in 4b2c8c02d83ba54b7eb80957caa042321ba16ee2

File: src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h
Comment (GiulioRomualdi) [reply src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h:59]: As written in [stackoverflow](https://stackoverflow.com/questions/21132538/correct-usage-of-the-eigenref-class)
> Then, there is not much difference between a `const Ref<const T>` and a `const Ref<const T>&` because it is unlikely that the function is called with an existing `Ref<const T>` object, and so a `Ref<const T>` will have to be created in most cases.

Since I don't call the functions on an existing `Ref` object. I choose:
-  `Eigen::Ref<const Eigen::VectorXd>` for readable only object
-  `Eigen::Ref<Eigen::VectorXd>` for writable object

I like the symmetry :smile: 

@traversaro @S-Dafarra what do you think about this?

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: Done in https://github.com/dic-iit/bipedal-locomotion-framework/pull/83/commits/121df8a8cef83ef427ef8d5da522f60ac2475c33

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: Done in https://github.com/dic-iit/bipedal-locomotion-framework/pull/83/commits/121df8a8cef83ef427ef8d5da522f60ac2475c33

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp:267]: what do you think about this? 

```suggestion
        if (std::abs(polynomials[i].duration) <= std::numeric_limits<double>::epsilon())
```

File: src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h
Comment (S-Dafarra) [reply src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h:59]: I read ``Eigen::Ref`` as a fancy substitution for ``&``. In addition, it should be a very lightweight class with no dynamic memory allocation. I prefer passing them by copy, so I agree with your choice.

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp:584]: Well you are actually right the computation is a bit cryptic

given a generic internal point, the submatrix A is given by 
image 
this submatrix has 12 entries different from zero (i.e. 12 triplets)

for the first and the last points, the submatrix A contains only 8 entries. Indeed, in the case of the first internal point, the submatrix A does not contain the first two columns, and vice-versa in the case of the last internal point,  the submatrix A does not contain the last two columns.

On the light of this the `numberOfExpectedTriplets` is equal to
`numberOfExpectedTriplets = 12 * (numberOfInteriorKnots - 2) + 8 * 2`
I can update the code in this way

```suggestion
        constexpr int numberOfNonZeroExtremalPoints = 8;
        constexpr int numberOfNonZeroInternalPoints = 12;
        constexpr int extremalPoints = 2;
        numberOfExpectedTriplets = numberOfNonZeroInternalPoints * (numberOfInteriorKnots - extremalPoints)  + numberOfNonZeroExtremalPoints * extremalPoints;
```

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: Done in 97ebb0bfe8ea740c1318b2601865ae869e0ca41c

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: There was a typo in the issue comment. Fixed! Thanks for pointing out.

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: Done in https://github.com/dic-iit/bipedal-locomotion-framework/pull/83/commits/97ebb0bfe8ea740c1318b2601865ae869e0ca41c

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp:499]: Yep, `addKnownTermKnotPosition` is called only in this part of the code and here I reset to zero the vector every time 
https://github.com/dic-iit/bipedal-locomotion-framework/blob/97ebb0bfe8ea740c1318b2601865ae869e0ca41c/src/Planners/src/QuinticSpline.cpp#L618-L644

File: src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h
Comment (GiulioRomualdi) [reply src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h:23]: > I don't think these things are mutually exclusive. Inheriting from Advanceable would imply having two additional methods, but you can still use it as you prefer.

You are right, the only thing I don't like of advanceable in this case is the fact that you have to set the sampling time of the class (that could be not required) furthermore it is not really clear how the user should use it.
In other words:
```cpp
QuinticSpline sp;
// set the spline
sp.setSamplingTime(0.01)

auto traj = sp.get(); // <-------- Here I get the first value of the trajectory ( at t =  0 )
sp.advance(); // <----- Advace the trajectory
auto traj = sp.get(); // <-------- Here I get the second value of the trajectory ( at t = 0.01 )

sp.evaluatePoint(10, pos, vel, acc); // <-------- pos, vel and acc contains the state at t = 10

sp.advance(); // <----- Advace the trajectory
auto traj = sp.get(); // <-------- Here I get the second value of the trajectory ( at t = 0.02 )
```
I don't know if this may create problems in the future. If you think no I can add the possibility to set the sampling time. If it is set then the advance machinery works otherwise the user can get the trajectory only with `evaluatePoint()`.

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: Done here: 53dc2b22795518d8e3f00ebe1a6ed4887fcc30ec

File: src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h
Comment (S-Dafarra) [reply src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h:23]: > You are right, the only thing I don't like of advanceable in this case is the fact that you have to set the sampling time of the class (that could be not required)

That is true. We can define a method like ``setAdvanceTimeStep``. In this way, it is clear that we are setting the time step for the ``advance`` method. We can discuss if this should be part of ``Advanceable`` or not. Regarding the example you showed, I don't think it is a big issue. ``Advanceable`` gives a basic interface for all the objects which output a trajectory. Then, if you have a reference to the actual class you can use its full-fledged interface.

File: src/Planners/src/QuinticSpline.cpp
Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp:596]: I think we can simply name them `numberOfLHSExtremumElements` and `numberOfLHSNonExtremumElements` which are 4 and 6 respectively, which gives,

```cpp
numberOfExpectedTriplets = 2*(numberOfLHSNonExtremumElements*(numberOfInteriorKnots - numberOfExtremum) + 2 * numberOfLHSExtremumElements )
```

What do you think?

File: src/Planners/src/QuinticSpline.cpp
Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp:279]: Yes. I think this should be fine.

File: src/Planners/src/QuinticSpline.cpp
Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp:511]: Understood.

File: src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h
Comment (GiulioRomualdi) [reply src/Planners/include/BipedalLocomotion/Planners/QuinticSpline.h]: https://github.com/dic-iit/bipedal-locomotion-framework/pull/83/commits/02557d7c8b4ccfef3bc6d147cb4aa00283110274 introduces the advance capabilities https://github.com/dic-iit/bipedal-locomotion-framework/pull/83/commits/5eb51cdf3cb1de6aed87548c5ce0224ed07b3612 is the test.

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp]: Done in 15982eaed361916a898f8884eabc138bd1b445ed

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp:623]: I don't know if `LHS` is a good acronym in this case. Is it related to the left-hand side of the aforementioned equation?

Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp:623]: > Is it related to the left-hand side of the aforementioned equation?

Yes.

File: src/Planners/src/QuinticSpline.cpp
Comment (S-Dafarra) [inline src/Planners/src/QuinticSpline.cpp:805]: Notice that ``find_if`` performs a linear search, not a binary search. In any case, this is not a big deal with few knot points.

**Review by S-Dafarra — APPROVED** [review]

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp:623]: > > Is it related to the left-hand side of the aforementioned equation?
> 
> Yes.

Ok because in that part of the code it is not easy to understand what is the right-hand side of the equation

File: src/Planners/src/QuinticSpline.cpp
Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp:805]: We should keep in mind this if we notice performance limitations.

File: src/Planners/src/QuinticSpline.cpp
Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp:623]: Then I think just the distinction between extremum and nonExtremum elements would be sufficient.

Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp:623]: Because mentioning them as,
```
constexpr int numberOfNonZeroExtremalPoints = 8;
constexpr int numberOfNonZeroInternalPoints = 12;
```

is a bit confusing as well.

Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp:623]:
```cpp
constexpr int numberOfExtremumElement = 8;
constexpr int numberOfNonExtremumElement = 12;
```

what do you think?

Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp:623]: What I meant was, 
since the number of columns is 4 or 6 depending on if its an extremum or a non-extremum, we say 
```
constexpr int numberOfExtremumElements = 4;
constexpr int numberOfNonExtremumElements= 6;
```

and then we compute the `numberOfExpectedTriplets` corresponding to `x` and `xdot` as,

```
numberOfExpectedTriplets = 2*(numberOfNonExtremumElements*(numberOfInteriorKnots - numberOfExtremum) + 2 * numberOfExtremumElements )
```

Sorry for being picky here. Do you think this will be a more readable mapping of the expected matrix?

Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp:623]: Or do you think it's more meaningful to define them as in https://github.com/dic-iit/bipedal-locomotion-framework/pull/83#discussion_r482871832

Comment (GiulioRomualdi) [reply src/Planners/src/QuinticSpline.cpp:623]: > ```
> fNonExtremumElements*(numberOfInteriorKnots - numberOfExtremum) + 2 * numberOfExtremumElements )
> ```
> 
> 
> Sorry for being picky here. Do you think this will be a more readable m

This is fine. I'm working on this

Comment (S-Dafarra) [reply src/Planners/src/QuinticSpline.cpp]: I love the A :joy:

Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp]: Lovely documentation!!! Respect :rocket:

More love for the A :heart:

Comment (prashanthr05) [reply src/Planners/src/QuinticSpline.cpp]: Minor changes https://github.com/dic-iit/bipedal-locomotion-framework/commit/7d163f19c9ec8af0a3b413240f1bfa4605c66d3b#r42012518

**Review by prashanthr05 — APPROVED** [review]

**Comment by prashanthr05** [comment]
I have given my approval for the PR.

However, please fix the typoes before merging, :)

**Pull request merged**
